packages/common/infra/src/sync/indexer/README.md

```diff
@@ -0,0 +1,147 @@
+# index
+
+Search engine abstraction layer for AFFiNE.
+
+## Using
+
+1. Define schema
+
+First, we need to define the shape of the data. Currently, there are the following data types.
+
+- 'Integer'
+- 'Boolean'
+- 'FullText': for full-text search, it will be tokenized and stemmed.
+- 'String': for exact match search, e.g. tags, ids.
+
+```typescript
+const schema = {
+  title: 'FullText',
+  tag: 'String',
+  size: 'Integer',
+} as Schema;
+```
+
+> **Array type**
+> All types can contain one or more values, so each field can store an array.
+
+2. Pick a backend
+
+Currently, there are two backends available.
+
+- `MemoryIndexer`: in-memory indexer, useful for testing.
+- `IndexedDBIndex`: persistent indexer using IndexedDB.
+
+> **Underlying Data Table**
+> Some back-end processes need to maintain underlying data tables, including table creation and migration. This operation should be silently executed the first time the indexer is invoked.
+> Callers do not need to worry about these details.
+>
+> This design conforms to the usual conventions of search engine APIs, such as in Elasticsearch: https://www.elastic.co/guide/en/elasticsearch/reference/current/array.html
+
+3. Write data
+
+Write data to the indexer. you need to start a write transaction by `await index.write()` first and then complete the batch write through `await writer.commit()`.
+
+> **Transactional**
+> Typically, the indexer does not provide transactional guarantees; reliable locking logic needs to be implemented at a higher level.
+
+```typescript
+const indexer = new IndexedDBIndex(schema);
+
+const writer = await index.write();
```

Comment:
is there a higher level wrapper to hide the detail of typed indexer?

Reply:
No, is already a high-level abstraction